squirrel.frontier/src/main/java/org/dice_research/squirrel/data/uri/norm/NormalizerImpl.java

```diff
@@ -20,122 +24,218 @@
+     */
+    private final static BitSet UNESCAPED_CHARS = new BitSet(0x7F);
+
+    static {
+        /*
+         * https://tools.ietf.org/html/rfc3986#section-2.2 For consistency,
+         * percent-encoded octets in the ranges of ALPHA (%41-%5A and %61-%7A), DIGIT
+         * (%30-%39), hyphen (%2D), period (%2E), underscore (%5F), or tilde (%7E)
+         * should not be created by URI producers and, when found in a URI, should be
+         * decoded to their corresponding unreserved characters by URI normalizers.
+         */
+        UNESCAPED_CHARS.set(0x2D, 0x2E);
+        UNESCAPED_CHARS.set(0x30, 0x39);
+        UNESCAPED_CHARS.set(0x41, 0x5A);
+        UNESCAPED_CHARS.set(0x61, 0x7A);
+        UNESCAPED_CHARS.set(0x5F);
+        UNESCAPED_CHARS.set(0x7E);
+    }
+
+    /**
+     * map containing schemes and their default ports
+     */
+    private static final Map<String, Integer> defaultPortMap = new HashMap<>();
+    static {
+        defaultPortMap.put("http", 80);
+        defaultPortMap.put("https", 443);
+        defaultPortMap.put("ftp", 21);
+        defaultPortMap.put("ftps", 990);
+        defaultPortMap.put("sftp", 22);
+
+    }
+
+    /**
+     * list containing query parameters for session ids     *
+     */
+    private static final List<String> sessionIDs = new ArrayList<>();
+    static {
+        sessionIDs.add("sessionid");
+        sessionIDs.add("jsessionids");
+        sessionIDs.add("phpsessid");
+        sessionIDs.add("sid");
+    }
+
+    @Override
+    public CrawleableUri normalize(CrawleableUri uri) {
+        URI uriObject = uri.getUri();
+        boolean changed = false;
+        // normalize path
+        String path = uriObject.getRawPath();
+        if (path != null) {
+            String temp = normalizePath(path);
+            if (temp != path) {
+                path = temp;
+                changed = true;
+            }
+        }
+
+
+        // Copy Normalization from
+        // https://github.com/crawler-commons/crawler-commons/blob/master/src/main/java/crawlercommons/filters/basic/BasicURLNormalizer.java
+        // OR use URI.normalize()
+
+        // Check whether the query part of a URI has to be sorted
+        String query = uriObject.getQuery();
+        if(query != null){
+            if(query.length() > 0) {
+                String[] queryList = query.split("&");
+                Arrays.sort(queryList);
+                List<String> queries = new ArrayList<>(Arrays.asList(queryList));
+                List<String> toRemove = new ArrayList<>();
+                for(String queryParameter : queries){
+                    //removing session ids
+                    if(sessionIDs.contains(queryParameter.split("=")[0].toLowerCase())){
+                        toRemove.add(queryParameter);
+                    }
+                }
+                queries.removeAll(toRemove);
+                query = String.join("&", queries);
+                changed = true;
```

Comment:
Your code always states that something has been changed, although it is not sure whether this is the case. It would be better if you could check whether the part of the URI really changed :thinking: